Studio: User can't delete expired demo site from Studio

katinthehatsite opened the pull request:
Closes https://github.com/Automattic/dotcom-forge/issues/7082

## Proposed Changes

This PR adds `Clear expired site` button to the expired screen under `Share` tab in Studio:

<img width="1023" alt="Screenshot 2024-06-07 at 4 32 35 PM" src="https://github.com/Automattic/studio/assets/25575134/72ab0f71-286b-4481-9ead-944cdb0ab3a4">

## Testing Instructions

* Pull the changes from this PR 
* Start the app with `nvm use && npm install && npm start`
* Create a demo site
* Force the demo site to be expired by changing the [following line](https://github.com/Automattic/studio/blob/ffd73b9a60113ed55785ea03fed6aeaf999d937f/src/hooks/use-expiration-date.ts#L24) in the `use-expiration-date.ts`:

```suggestion
let isExpired = false;
```

to `true`:

```suggestion
let isExpired = true
```

* Observe `Clear expired site` button
* Click the button, observe a spinner and confirm that you are seeing the `<NoSnapshots>` component
* Try testing `Clear expired site` without internet connection and confirm that the button is disabled in this case

**Notes**

* There is a loading spinner present which was not mentioned in the initial design. I tried without it but I found the shift in the layout to be too abrupt. Happy to reconsider this approach
* Once you focus on the button to `Clear expired site`, you will see the borders appear. I am not sure if they should be there according to the design. Happy to remove those as well.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

Commit:
Add button and functionality to reset state when there is an expired site

Commit:
Apply styles to the button

Commit:
Adjust logic for resetting state

Commit:
Reuse logic for deleting snapshots

Commit:
Reuse deletion code

Commit:
Handle offline mode

Commit:
Remove an empty line

Commit:
Cleanup

wojtekn commented on src/components/content-tab-snapshots.tsx:
```diff
@@ -435,21 +457,36 @@ function AddDemoSiteWithProgress( {
+						<Tooltip disabled={ ! clearTooltipContent } { ...clearTooltipContent }>
+							<Button
+								aria-description={ clearTooltipContent?.text ?? '' }
+								disabled={ isOffline }
```
As it's about deleting a site from the local storage only, do we need to disable it in offline mode?

katinthehatsite replied:
The reason why I initially implemented it was because I re-used `deleteSnapshot` and it first makes an API request and then removes the snapshots stored locally so it would fail in the offline mode. Also, it seemed consistent in the UI with adding a demo site when the site was offline.

 On a second thought, I could probably use `removeSnapshot` directly instead and only target the locally stored snapshots in `appdata-v1.json`.

I will re-test now with the approach you mentioned for testing.

wojtekn commented:
@katinthehatsite  I tried to test using steps closer to real-world scenarios - added one of my expired demo sites manually in the `appdata-v1.json` file and launched Studio. The demo site wasn't displayed at all. I tried to remove demo site using API endpoint directly, restarted Studio, and it disappeared, too. Maybe this is no longer an issue, and we don't need this button anymore?

It happens because we introduced the cleanup feature via `clearFloatingSnapshots` at https://github.com/Automattic/studio/pull/131.

katinthehatsite commented:
> I tried to test using steps closer to real-world scenarios - added one of my expired demo sites manually in the appdata-v1.json file and launched Studio. The demo site wasn't displayed at all.

@wojtekn I was trying to test this but I am not completely sure how you did this step? Did you force the site to somehow expire first (e.g. by editing the date in `appdata-v1.json`) before testing or you already had a site that expired and was present in your snapshots in `appdata-v1.json`?

> It happens because we introduced the cleanup feature via `clearFloatingSnapshots` at https://github.com/Automattic/studio/pull/131.

It makes sense. I would still like to test in the way that you ran your test but if that's the case, should we still need the UI for the expired demo site view?

katinthehatsite commented:
@wojtekn I manually edited the `date` in the snapshot in the `appdata-v1.json` file but I am getting an expired view on my end:

<img width="665" alt="Capture d’écran, le 2024-06-10 à 20 32 23" src="https://github.com/Automattic/studio/assets/25575134/08c5a198-0cea-4c89-a4d3-8f6f32261769">

How did you trigger the site to be expired exactly? Thanks!

wojtekn commented:
> @wojtekn I was trying to test this but I am not completely sure how you did this step? Did you force the site to somehow expire first (e.g. by editing the date in appdata-v1.json) before testing or you already had a site that expired and was present in your snapshots in appdata-v1.json?

I deleted it using `/jurassic-ninja/delete` endpoint on https://developer.wordpress.com/docs/api/console/.

> It makes sense. I would still like to test in the way that you ran your test but if that's the case, should we still need the UI for the expired demo site view?

Yes, this is what I meant in my comment above:

> Maybe this is no longer an issue, and we don't need this button anymore?

We should test this flow, and we could close the PR if it's no longer an issue.

jeroenpf commented:
@katinthehatsite - What I observed is that a deleted/expired site is automatically deleted from the app when you restart it. I think this change is still valid because if you don't restart the app, the demo site will still be there.

However, perhaps one improvement could be to not rely on a hardcoded expiry but instead poll the status endpoint for a site once in a while when the 'share' tab is active. If a site is deleted/expired we could automatically delete it from the local storage, or perhaps show something in the UI to let the user know their previous demo site has expired? In any case, I'm not sure if its a great experience if a demo site suddenly disappears without any feedback to the user that this happened.

katinthehatsite commented:
I tried some testing today:

* I created a new site, tried deleting the site from the endpoint and then restarting the app -> it went back to the normal screen to add a new site
* I created a new site, tried deleting the site from the endpoint and then moved around tabs etc. -> the site was still present
* I created a new site, updated `isExpired` to true, deleted the site from the endpoint, saw the expired screen, moved around tabs etc. -> the expired view was present
* I created a new site, updated `isExpired` to true, deleted the site from the endpoint, saw the expired screen, restarted the app  -> the screen went back to the usual screen to add a site

So I think @jeroenpf is right that the screen would stay until the app restarts. I don't have a strong opinion on how to proceed:
* We can leave the screen without adding any further button since it would be cleared when the app restarts and if the site expires
* We can proceed with the implementation to allow the users who don't restart the app often to clear the site. I would need to refactor a bit the code and remove offline mode handling

Personally, I think it is something that is nice to have but I don't find it to be an important feature because it seems to be a bit of an edge case now that we have `clearFloatingSnapshots`. I am happy to proceed either way

wojtekn commented:
> In any case, I'm not sure if its a great experience if a demo site suddenly disappears without any feedback to the user that this happened.

This is a good point, @jeroenpf . What if we go with @katinthehatsite's solution and remove `clearFloatingSnapshots()`? That way, snapshots wouldn't disappear automatically. We would likely need to change how we delete those snapshots to avoid trying to call API endpoint, though.

katinthehatsite commented:
Sounds good @wojtekn - I will adjust the solution based on that and re-request a review again 👍

Commit:
Refactor the snapshots deletion

Commit:
Fix linter

Commit:
Remove logic to clear floating snapshots

Commit:
Remove logic related to clearing expired snapshots

Commit:
Add tests

Commit:
Improve test

Commit:
Improve test

katinthehatsite commented:
@wojtekn @jeroenpf Thanks for suggestions! I have now done the following:

* removed `clearFloatingSnapshots` to remove the cleanup when the app restarts
* used `removeSnapshot` instead of `deleteSnapshot` as it does not call the API and only takes care of the local snapshot
* added some tests

One thing that I have not added is the dialog that asks if the user wants to confirm the clearing of the expired demo site. On one hand, it would make the experience with other destructive buttons in the app and make the UI shift a bit less jarring. On the other hand, it might be a bit of an overkill. Happy to add it if you think we should do it.

Commit:
Clean up empty lines and improve test description

fluiddot commented on src/components/content-tab-snapshots.tsx:
```diff
@@ -150,6 +151,8 @@ function SnapshotRow( {
 						selectedSite={ selectedSite }
 						isSnapshotLoading={ snapshot.isLoading }
 						tagline={ __( "We're creating your new demo site." ) }
+						isExpired
```
I wonder if the Clear expired site button should be rendered at the same level as `AddDemoSiteWithProgress` as both are buttons. WDYT?

katinthehatsite replied:
I think it makes sense - I made the adjustment here: https://github.com/Automattic/studio/pull/218/commits/7235e394104ae6aadd18cca95d3384d5ae0b9d14

fluiddot reviewed (commented):
I noticed the following issue when testing:

**Steps:**
- Add a demo site.
- Change the date of the site to `0` in `appdata-v1.json` file.
- Observe the site is marked as expired.
- Click on Add demo site.
- Wait until the demo site is created.
- Delete the demo site.
- Observe the previously expired site is shown again.

I understand that an expired site should be removed when a new demo site is created.

katinthehatsite commented:
Hey @fluiddot thanks for testing! I tried these steps but I am not seeing the expired site on my end once I click on `Add site`. Or are you referring to the expired site being in `appdata-v1.json`?

I compared the behavior to trunk and I see the exactly same experience there as well. 

I am wondering though if we should completely remove clearing floating snapshots or leave it for backup. I like having some sort of cleanup job being present.

fluiddot commented:
> Hey @fluiddot thanks for testing! I tried these steps but I am not seeing the expired site on my end once I click on `Add site`. Or are you referring to the expired site being in `appdata-v1.json`?

The expired demo site is displayed after deleting the demo site created after the demo site expires.

> I compared the behavior to trunk and I see the exactly same experience there as well.

Ah, true. I've just checked `trunk` and noticed that the expired snapshot shows up too. Hence, I wouldn't tackle this problem in this PR.

> I am wondering though if we should completely remove clearing floating snapshots or leave it for backup. I like having some sort of cleanup job being present.

Good question. I presume that expired demo sites won't be useful for users as they can't access them. So, I think there's no need to keep them as a backup.

fluiddot commented on src/components/tests/content-tab-snapshots.test.tsx:
```diff
@@ -474,4 +474,41 @@ describe( 'AddDemoSiteWithProgress', () => {
 		expect( offlineTooltip ).toBeVisible();
 	} );
+
+	test( 'confirms that Clear expired site button is present on expired snapshot and calls removeSnapshot when clicked', async () => {
```
I wonder if we could also cover in this test case that the demo site is not shown after removing the snapshot. WDYT?

katinthehatsite replied:
Makes sense! Let's add it

katinthehatsite replied:
I made some updates for it here: https://github.com/Automattic/studio/pull/218/commits/678d9068aa86f39151e9b891981b2d634648e77b

Commit:
Extend test to cover the snapshot removal

Commit:
Add back clear floating snapshots job

katinthehatsite commented:
After discussing this with @kozer who initially created `clearFloatingSnaphots` function, I think that we should keep it. Hence, I have done the following:

* added back `clearFloatingSnapshots` so that it gets executed when the app restarts if the user has not cleared the snapshot manually. I don't think the change in the UI is too radical for the user for the expired site to be removed when the app restarts so this seems acceptable to me;
* kept the button to `Clear expired site` to have an option for the user to manually remove the expired snapshot

Commit:
Move the button rendering level

Commit:
Move tests outside of add demo sites with progress

wojtekn reviewed (approved):
Thanks @katinthehatsite, let's ship it.

If we see that users are confused by the fact that an expired site or a site removed by support disappears automatically, we may reassess and potentially remove the auto-clear logic and update the wording.

fluiddot commented on src/components/content-tab-snapshots.tsx:
```diff
@@ -151,6 +152,13 @@ function SnapshotRow( {
 						tagline={ __( "We're creating your new demo site." ) }
 					/>
+					<Button
+						aria-description={ __( 'Clear expired site' ) }
```
This won't be needed as screen readers take the description based on the button's label, which in this case is the same string "Clear expired site".

katinthehatsite replied:
Updated in https://github.com/Automattic/studio/pull/218/commits/10e71e3075aff4dec6279fc102d08e4d51802a4f

fluiddot commented on src/components/content-tab-snapshots.tsx:
```diff
@@ -379,6 +387,8 @@ function AddDemoSiteWithProgress( {
 	selectedSite: SiteDetails;
 	className?: string;
 	tagline?: string;
+	isExpired?: boolean;
+	onClearExpiredSite?: () => void;
```
These two props seem not to be included in the props list of `AddDemoSiteWithProgress` component so we could remove them.

katinthehatsite replied:
Removed in https://github.com/Automattic/studio/pull/218/commits/32f5d9fc3fd4f72b209e61ad703bc2d6a0b8b4f2

nice catch! I was moving the button to be rendered on the same level as `addDemoSiteWithProgress` and these were the leftovers ✅

fluiddot reviewed (commented):
By modifying the date of a previously created snapshot in the `appdata-v1.json` file, I encountered that adding a demo site results in a state where it shows both the loading indicator and the "Clear expired site" button. This can result in an unexpected behavior when clicking "Clear expired site" button (see attached video capture):

> [!NOTE]
> Clicking on the "Clear expired site" button right after adding a demo site clears the previous snapshot properly. To reproduce the issue, you need to wait a bit until the snapshot is created (i.e. it's uploaded) but still in the loading state until the API confirms the site is ready.

https://github.com/user-attachments/assets/4b46b5b7-db5c-45f9-b068-dd88ea6cbbd9

Commit:
Remove unused props

Commit:
Remove extra aria label

Commit:
Fix issue with progress bar displayed along the progress bar

katinthehatsite commented:
Thanks @fluiddot - I have made the adjustments here to fix the issue with the button display: https://github.com/Automattic/studio/pull/218/commits/1fa0063c3bdc8caf7e4e94f745eac374ecb2f148

Commit:
Resolve conflicts with trunk

fluiddot commented:
> Thanks @fluiddot - I have made the adjustments here to fix the issue with the button display: [1fa0063](https://github.com/Automattic/studio/commit/1fa0063c3bdc8caf7e4e94f745eac374ecb2f148)

I've tested `trunk` after this PR got merged and confirmed it works as expected. Sorry for the delay in the review 🙇 .
